File: 프로그래머스/unrated/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    l = len(friends)
    name = {friends[x]:x for x in range(l)}
    graph = [[0 for _ in range(l)] for _ in range(l)]
    score = [0 for _ in range(l)]
    for i in range(len(gifts)):
        a, b = gifts[i].split()
        graph[name[a]][name[b]] += 1
        score[name[a]] += 1
        score[name[b]] -= 1

    answer = [0 for _ in range(l)]
    for r in range(l):
        for c in range(r+1, l):
            a, b = graph[r][c], graph[c][r]
            if a == b:
                if score[r] > score[c]:
                    answer[r] += 1
                elif score[r] < score[c]:
                    answer[c] += 1
            elif a > b:
                answer[r] += 1
            else:
                answer[c] += 1

    return max(answer)

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    
    public int solution(String[] friends, String[] gifts) {
        String[] s;
        int ans = 0;
        int tmp = 0;
        Map <String, Integer> index = new HashMap();
        for (int i = 0; i < friends.length; i++) index.put(friends[i], i);
        int[] score_array = new int[friends.length];
        int[][] gift_array = new int[friends.length][friends.length];
        for (int i = 0; i < gifts.length; i++){
            s = gifts[i].split(" ");
            gift_array[index.get(s[0])][index.get(s[1])] += 1;
            score_array[index.get(s[0])]+=1;
            score_array[index.get(s[1])]-=1;
        }
        for (int i = 0; i < gift_array.length; i++){
            tmp = 0;
            for (int j = 0; j < gift_array.length; j++) {
                if (gift_array[i][j] > gift_array[j][i]) tmp += 1;
                else if(gift_array[i][j] == gift_array[j][i] && score_array[i] > score_array[j]) tmp+=1;}
            ans = Math.max(ans, tmp);
        }
        return ans;
    }
}